File: 0001-two-sum/1.two-sum.cs
/*
 * @lc app=leetcode id=1 lang=csharp
 *
 * [1] Two Sum
 * 
 * Given an array of integers nums and an integer target, return indices of the two numbers such that they add up to target.
 * You may assume that each input would have exactly one solution, and you may not use the same element twice.
 * You can return the answer in any order.
 * 
 * Example:
 * Input: nums = [2,7,11,15], target = 9
 * Output: [0,1]
 * Explanation: Because nums[0] + nums[1] == 9, we return [0, 1].
 * 
 */

// @lc code=start
public class Solution {
    public int[] TwoSum(int[] nums, int target) {
        // Dictionary<value, index>
        Dictionary<int, int> numDict = new Dictionary<int, int>();
        int complement;

        for (int x = 0; x < nums.Length; x++)
        {
            if (!numDict.ContainsKey(nums[x]))
            {
                numDict.Add(nums[x], x);
                complement = target - nums[x];

                if (numDict.ContainsKey(complement))
                {
                    // if the complement is not the value just added, we are done
                    if (numDict[complement] != x)
                    {
                        return new int[2] {numDict[complement], x};
                    }
                    // else check if complement exists elsewhere
                    else if (Array.IndexOf(nums[(x+1)..], complement) > -1)
                    {
                        return new int[2] {x, Array.IndexOf(nums[(x+1)..], complement) + (x + 1)};
                    }
                }
            }
        }
        return null;
    }
}
// @lc code=end

File: 0001-two-sum/1.two-sum.py
#
# @lc app=leetcode id=1 lang=python3
#
# [1] Two Sum
#

# @lc code=start
from re import I


class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        numDict = dict()

        x = 0
        while x < len(nums):
            if nums[x] not in numDict:
                numDict[nums[x]] = x
                complement = target - nums[x]

                if complement in numDict:
                    # if the complement is not the value just added, we are done
                    if numDict[complement] != x:
                        return [numDict[complement], x]
                    # else check if the complement exists elsewhere
                    elif complement in nums[x+1:]:
                        return [x, nums[x+1:].index(complement) + (x + 1)]

            x += 1

        return
# @lc code=end

